### PR TITLE
Refs[TASK][34793]Create sqlite to save favourite meal

### PR DESCRIPTION
**Pull request by GiangGiang1105**
## Related Tickets
- [#Task 34793:Create sqlite to save favourite meal](https://edu-redmine.sun-asterisk.vn/issues/34793)
## WHAT
- Create sqlite to save favourite meal
## Review Checklist

Category | View Point | Description | Expected Reviewer Answer | Self review | Reviewer2 (name)
--- | --- | --- | --- | --- | ---
Conventions | Does the code follow Sun* coding style and coding conventions? | https://github.com/framgia/coding-standards/tree/master/eng/android | YES |<li>- [ ] yes</li>|<li>- [ ] yes</li>  
Redmine | Does the ticket follow Sun* Redmine working process?  | https://github.com/framgia/Training-Guideline/blob/master/WorkingProcess/redmine/redmine.md| YES |<li>- [ ] yes</li>|<li>- [ ] yes</li>  
Documentation | Is there any incomplete code? If so, should it be removed or flagged with a suitable marker like ‘TODO’? |  | YES |<li>- [ ] yes</li>|<li>- [ ] yes</li>  
## Notes (Optional)
*(Impacted Areas in Application(List features, api, models or services that this PR will affect))*

*(List gem, library third party add new)*

*(Other notes)*

### CODE REVIEW

File: app/src/main/java/com/example/gfood/screen/mealdetail/MealDetailPresenter.kt [hunk 1]
```diff
@@ -1,25 +1,81 @@
...
+import com.example.gfood.data.source.MealFavouriteDataSource
+import com.example.gfood.data.source.local.OnFetchDataLocalListener
 import com.example.gfood.data.source.remote.OnFetchDataJsonListener
 import com.example.gfood.data.source.repository.MealDetailRepository
 
-class MealDetailPresenter(private val repository: MealDetailRepository) :
-    ContractMealDetail.Presenter {
+class MealDetailPresenter(
+    private val repository: MealDetailRepository,
+    private val mealFavouriteLocalDataSource: MealFavouriteDataSource.Local
```
Comment (tuanlvt-1117) [inline app/src/main/java/com/example/gfood/screen/mealdetail/MealDetailPresenter.kt]: datasource sao lại ở trong presenter?

File: app/src/main/java/com/example/gfood/data/source/local/database/MealFavouriteDao.kt [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+    fun delete(
+        idEntity: String,
+        listener: OnFetchDataLocalListener<Int>
+    )
+
+    fun getMeal(
+        idEntity: String,
+        listener: OnFetchDataLocalListener<Boolean>
+    )
+}
```
Comment (tuanlvt-1117) [inline app/src/main/java/com/example/gfood/data/source/local/database/MealFavouriteDao.kt]: lặp code

File: app/src/main/java/com/example/gfood/data/source/local/database/dao/MealFavouriteDaoImpl.kt [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+            contentValue
+        ).also {
+            isSuccess = it
+        }
+        isSuccess?.let {
+            if (it > 0) {
+                listener.onSuccess(it)
+            } else {
+                listener.onFailed(R.string.failed_to_save)
+            }
```
Comment (tuanlvt-1117) [inline app/src/main/java/com/example/gfood/data/source/local/database/dao/MealFavouriteDaoImpl.kt]:
```suggestio
            if (it > 0) listener.onSuccess(it)
            else listener.onFailed(R.string.failed_to_save)
```

File: app/src/main/java/com/example/gfood/data/source/local/database/dao/MealFavouriteDaoImpl.kt [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+                            title = getString(getColumnIndex(MealFavouriteTable.COLUMN_NAME)),
+                            image = getString(getColumnIndex(MealFavouriteTable.COLUMN_IMAGE))
+                        )
+                    )
+                }
+            }
+            listener.onSuccess(meals)
+        } else {
+            listener.onFailed(R.string.failed_to_select)
+        }
```
Comment (tuanlvt-1117) [inline app/src/main/java/com/example/gfood/data/source/local/database/dao/MealFavouriteDaoImpl.kt]:
```suggestion
        } else listener.onFailed(R.string.failed_to_select)
```

File: app/src/main/java/com/example/gfood/data/source/local/database/dao/MealFavouriteDaoImpl.kt [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+        databaseHelper.writableDatabase.delete(
+            MealFavouriteTable.TABLE_NAME,
+            "${MealFavouriteTable.COLUMN_ID_MEAL}=?",
+            arrayOf(idEntity)
+        ).also {
+            if (it > 0) {
+                listener.onSuccess(it)
+            } else {
+                listener.onFailed(R.string.failed_to_delete)
+            }
```
Comment (tuanlvt-1117) [inline app/src/main/java/com/example/gfood/data/source/local/database/dao/MealFavouriteDaoImpl.kt]:
```suggestion
            if (it > 0) listener.onSuccess(it)
            else listener.onFailed(R.string.failed_to_delete)
```

File: app/src/main/java/com/example/gfood/data/source/local/database/dao/MealFavouriteDaoImpl.kt [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+        )
+        if (cursor != null) {
+            with(cursor) {
+                while (moveToNext()) {
+                    listener.onSuccess(true)
+                }
+            }
+        } else {
+            listener.onSuccess(false)
+        }
```
Comment (tuanlvt-1117) [inline app/src/main/java/com/example/gfood/data/source/local/database/dao/MealFavouriteDaoImpl.kt]:
```suggestion
        } else listener.onSuccess(false)
```

File: app/src/main/java/com/example/gfood/screen/mealdetail/MealDetailPresenter.kt [hunk 7]
```diff
@@ -1,25 +1,81 @@
...
+                view?.onGetMealDetailSuccess(data)
+            }
 
-                override fun onError(exception: Exception?) {
-                    view?.onError(exception!!)
+            override fun onError(exception: Exception?) {
+                view?.onError(exception!!)
+            }
+        }
+        )
```
Comment (tuanlvt-1117) [inline app/src/main/java/com/example/gfood/screen/mealdetail/MealDetailPresenter.kt]:
```suggestion
        })
```

File: app/src/main/java/com/example/gfood/screen/mealdetail/MealDetailPresenter.kt [hunk 8]
```diff
@@ -1,25 +1,81 @@
...
-                }
+        repository.getMealDetail(id, object : OnFetchDataJsonListener<MutableList<MealDetail>> {
+            override fun onSuccess(data: MutableList<MealDetail>) {
+                view?.onGetMealDetailSuccess(data)
+            }
 
-                override fun onError(exception: Exception?) {
-                    view?.onError(exception!!)
+            override fun onError(exception: Exception?) {
+                view?.onError(exception!!)
```
Comment (tuanlvt-1117) [inline app/src/main/java/com/example/gfood/screen/mealdetail/MealDetailPresenter.kt]: check nullable

File: app/src/main/java/com/example/gfood/screen/mealdetail/MealDetailPresenter.kt [hunk 9]
```diff
@@ -1,25 +1,81 @@
...
+            OnFetchDataLocalListener<MutableList<Meal>> {
+            override fun onSuccess(data: MutableList<Meal>) {
+                view?.onGetMealFavouriteSuccess(data)
+            }
+
+            override fun onFailed(idMessage: Int) {
+                view?.onFailed(idMessage)
+            }
+        }
+        )
```
Comment (tuanlvt-1117) [inline app/src/main/java/com/example/gfood/screen/mealdetail/MealDetailPresenter.kt]:
```suggestion
        })
```

File: app/src/main/java/com/example/gfood/screen/mealdetail/MealDetailPresenter.kt [hunk 10]
```diff
@@ -1,25 +1,81 @@
...
+        mealFavouriteLocalDataSource.save(meal!!, object : OnFetchDataLocalListener<Long> {
+            override fun onSuccess(data: Long) {
+                view?.onSaveMealFavourite()
+            }
+
+            override fun onFailed(idMessage: Int) {
+                view?.onFailed(idMessage)
+            }
+        }
+        )
```
Comment (tuanlvt-1117) [inline app/src/main/java/com/example/gfood/screen/mealdetail/MealDetailPresenter.kt]:
```suggestion
        })
```

File: app/src/main/java/com/example/gfood/screen/mealdetail/MealDetailPresenter.kt [hunk 11]
```diff
@@ -1,25 +1,81 @@
...
+    }
+
+    override fun onDeleteMealFavourite(idMeal: String?) {
+        mealFavouriteLocalDataSource.delete(idMeal!!, object : OnFetchDataLocalListener<Int> {
```
Comment (tuanlvt-1117) [inline app/src/main/java/com/example/gfood/screen/mealdetail/MealDetailPresenter.kt]: check nullable

File: app/src/main/java/com/example/gfood/screen/mealdetail/MealDetailPresenter.kt [hunk 12]
```diff
@@ -1,25 +1,81 @@
...
+
+            override fun onFailed(idMessage: Int) {
+                view?.onFailed(idMessage)
+            }
+        }
+        )
+    }
+
+    override fun onSaveMealFavourite(meal: Meal?) {
+        mealFavouriteLocalDataSource.save(meal!!, object : OnFetchDataLocalListener<Long> {
```
Comment (tuanlvt-1117) [inline app/src/main/java/com/example/gfood/screen/mealdetail/MealDetailPresenter.kt]: check nullable

File: app/src/main/java/com/example/gfood/screen/mealdetail/MealDetailPresenter.kt [hunk 13]
```diff
@@ -1,25 +1,81 @@
...
+            override fun onFailed(idMessage: Int) {
+                view?.onFailed(idMessage)
+            }
+        }
+        )
+    }
+
+    override fun onGetMeal(idEntity: String?) {
+        mealFavouriteLocalDataSource.getMeal(
+            idEntity!!,
```
Comment (tuanlvt-1117) [inline app/src/main/java/com/example/gfood/screen/mealdetail/MealDetailPresenter.kt]: check nullable

File: app/src/main/java/com/example/gfood/data/source/local/database/MealFavouriteDao.kt (hunk 2 above)
Comment (GiangGiang1105) [reply app/src/main/java/com/example/gfood/data/source/local/database/MealFavouriteDao.kt]: chỗ này lặp code là sao vậy anh? mỗi hàm chức năng riêng mà ạ

**Review by tuanlvt-1117 — APPROVED** [review]

**Pull request merged**